### PR TITLE
Implement arbitrary_enum_discriminant

### PR DESCRIPTION
**Pull request by jswrenn**
Implements RFC rust-lang/rfcs#2363 (tracking issue #60553).

### CODE REVIEW

File: src/libsyntax/feature_gate.rs
Comment (jswrenn) [inline src/libsyntax/feature_gate.rs]: It would be nice to `span_label` the individual unit variants here (as the original parse error did), but the `gate_feature` macros all immediately emit the diagnostic they generate.

File: src/libsyntax/feature_gate.rs
Comment (jswrenn) [reply src/libsyntax/feature_gate.rs]: I wasn't sure what the correct procedure is here. What version number should appear?

Comment (lzutao) [reply src/libsyntax/feature_gate.rs]: `1.36.0` is nightly now. So I think version should be `1.36.0`.

**Comment by RalfJung** [comment]
Would be nice to get a test case that has some CTFE code involving the new discriminants, just to make sure that works, too.

Cc @oli-obk do you know if this needs adjustments in Miri (specifically in `read_discriminant` and `write_discriminant`)?

**Comment by oli-obk** [comment]
>  do you know if this needs adjustments in Miri (specifically in read_discriminant and write_discriminant)?

No, this requires no changes for anything beyond the AST. Everything else is already setup to handle this.

Some const eval tests would be nice though.

File: src/libsyntax/feature_gate.rs
Comment (pnkfelix) [reply src/libsyntax/feature_gate.rs]: yes this sounds right to me

File: src/libsyntax/parse/parser.rs
Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: This `if-else` chain can be put in a `let struct_def = ...;` now, to simplify it.

**Comment by oli-obk** [comment]
So... there's this fun comment in the layouting source: https://github.com/rust-lang/rust/blob/a9ec99f4201ec33026a468ef1289f98a95b4d71a/src/librustc/ty/layout.rs#L879-L883

I'm guessing you can observe this as follows:

```rust
enum Foo {
    A(&'static i32),
    B,
}
enum Bar {
    A(&'static i32) = 0,
    B = 1,
}
assert_eq!(std::mem::size_of::<Foo>() == std::mem::size_of::<usize>());
assert_eq!(std::mem::size_of::<Bar>() > std::mem::size_of::<usize>());
```

**Comment by jswrenn** [comment]
@Ralph

> Would be nice to get a test case that has some CTFE code involving the new discriminants, just to make sure that works, too.

Would something like this fit the bill?
```rust
#![feature(arbitrary_enum_discriminant,const_raw_ptr_deref)]

#[allow(dead_code)]
#[repr(u8)]
enum Enum {
    Unit = 3,
    Tuple(u16) = 2,
    Struct {
        a: u8,
        b: u16,
    } = 1,
}

impl Enum {
    const unsafe fn tag(&self) -> u8 {
        *(self as *const Self as *const u8)
    }
}

const UNIT_TAG:   u8 = unsafe { Enum::Unit.tag() };
const TUPLE_TAG:  u8 = unsafe { Enum::Tuple(5).tag() };
const STRUCT_TAG: u8 = unsafe { Enum::Struct{a: 7, b: 11}.tag() };

fn main() {
    assert_eq!(3, UNIT_TAG);
    assert_eq!(2, TUPLE_TAG);
    assert_eq!(1, STRUCT_TAG);
}
```

**Comment by jswrenn** [comment]
@oli-obk To be sure I understand correctly: is the expected behavior that both `Foo` and `Bar` (being `#[repr(Rust)]`) should have the same size in that example?

**Comment by oli-obk** [comment]
No, it is perfectly fine for them to have different sizes for now, it's just an implementation detail that we can fix in the future. Maybe add my code as a canary test?

One thing that I'm wondering: the RFC states that the only legal way to get the discriminant back is to start out with a `repr(some_int)` enum. Would it make sense to forbid arbitrary enum discriminants on other `repr` enums for now?

One thing that I think should also have a test, too: code matching on an enum with both fields and explicit discriminants.

**Comment by RalfJung** [comment]
> Would something like this fit the bill?

Looks great!
Aren't there also safe conversion casts (`x as i32`)?

**Comment by jswrenn** [comment]
@oli-obk 

> One thing that I'm wondering: the RFC states that the only legal way to get the discriminant back is to start out with a `repr(some_int)` enum. Would it make sense to forbid arbitrary enum discriminants on other `repr` enums for now?

I think you're referring to [the stipulation](https://github.com/rust-lang/rfcs/blob/93f80305251c59dbf7289a2b32cca0819cafeb34/text/2363-arbitrary-enum-discriminant.md#semantics) that:
> enumerations with fields still can't be casted to numeric types with the as operator

As it stands, the functionality enabled by this PR forbids you from `as`-casting a variant of an enum if _any_ variant has data associated with it. For instance, this is _already_ forbidden:
```rust
#![feature(arbitrary_enum_discriminant)]

#[repr(u8)]
#[allow(dead_code)]
enum Enum {
  Unit = 3,
  Tuple(()) = 2,
  Struct {
    a: (),
    b: (),
  } = 1,
}

fn main() {
  println!("{:?}", Enum::Unit as u8);
}
```
and produces this error:
<details><summary>(Click to Expand)</summary>

```
error[E0605]: non-primitive cast: `Enum` as `u8`
  --> src/main.rs:21:20
   |
21 |   println!("{:?}", Enum::Unit as u8);
   |                    ^^^^^^^^^^^^^^^^
   |
   = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait
```

</details>

But this _is_ permitted:
```rust
#![feature(arbitrary_enum_discriminant)]

#[allow(dead_code)]
enum Enum {
  Unit = 3,
  Tuple() = 2,
  Struct {} = 1,
}

fn main() {
  assert_eq!(3, Enum::Unit as u8);
  assert_eq!(2, Enum::Tuple() as u8);
  assert_eq!(1, Enum::Struct{} as u8);
}
```
...and I think it's reasonable and consistent to permit that, since rust already permits writing:
```rust
#[allow(dead_code)]
enum Enum {
  Unit,
  Tuple(),
  Struct{},
}

fn main() {
  assert_eq!(0, Enum::Unit as u8);
  assert_eq!(1, Enum::Tuple() as u8);
  assert_eq!(2, Enum::Struct{} as u8);
}
```

I can't think of a compelling reason to add an explicit `repr` requirement given the above restrictions.

However, although the RFC explicitly does not enable this functionality, it seems to me like `as-casting` variants with fields (or something analogous) _should_ be permitted on enums with fields when an explicit `repr(inttype)` is provided. Since specifying a `repr(inttype)` enforces both the location and size of the discriminant, and the whole point of this RFC is to let the user explicitly enforce that the discriminant is a particular value, it _really_ seems like there should be a safe, ergonomic mechanism for getting that value back out—`unsafe { *(value as *const Self as *const u8) }` is quite a mouthful!

**Comment by oli-obk** [comment]
`unsafe { *(value as *const Self as *const u8) }` is only valid if you are using `repr(u8)`. This is mainly what I was referring to. Since the unsafe pointer magic is the only thing making it possible right now to extract he discriminant value, and it requires `repr(u8)`, I was wondering if we may as well require `repr(u8)` (or other ints) for enums with variants with fields and explicit discriminants.

File: src/test/run-pass/arbitrary_enum_discriminant.rs
Comment (RalfJung) [inline src/test/run-pass/arbitrary_enum_discriminant.rs]: This is not how CTFE works. The code here will run an run-time.

Only array lengths as well as `static` and `const` initializers are executed during CTFE (and things promoted to static).

File: src/test/run-pass/arbitrary_enum_discriminant.rs
Comment (RalfJung) [inline src/test/run-pass/arbitrary_enum_discriminant.rs]: Why `black_box`? That has no bearing on whether this is CTFE or not.

Comment (oli-obk) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: it does, because the code may get promoted otherwise. Due to the method call (`tag`) that doesn't happen though, but the `UNIT` may still get promoted.

Comment (RalfJung) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: The `UNIT` is a const, it doesn't need promotion.

Comment (oli-obk) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: Yea, there's no notable difference here

Comment (jswrenn) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: Since `UNIT` is const, and `Unit::tag` is a const fn, I worried that rustc (or a future rustc) could CTFE `UNIT.tag()`. Wrapping `UNIT` with `black_box` introduces a boundary between the value and the method call, hopefully preventing this (since this particular test is to ensure that discriminants are correct during _runtime_ execution).

Comment (Centril) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: Note that `black_box` is purely an optimization hint and not a semantic guarantee -- operationally you can only assume that it is the identity function.

Comment (RalfJung) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: Sure, but all we need here is that it is definitely not getting promoted statically.

**Comment by jswrenn** [comment]
@oli-obk 
> `unsafe { *(value as *const Self as *const u8) }` is only valid if you are using `repr(u8)`. This is mainly what I was referring to. Since the unsafe pointer magic is the only thing making it possible right now to extract he discriminant value, and it requires `repr(u8)`, I was wondering if we may as well require `repr(u8)` (or other ints) for enums with variants with fields and explicit discriminants.

I see your point. I think this is a good idea, and I'll take a stab at it in the next few days!

Implementation wise, I think [`rustc_typeck::check::check_enum`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/fn.check_enum.html) is the best place to add this check.

File: src/test/run-pass/arbitrary_enum_discriminant.rs
Comment (pnkfelix) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: I'm confused, has the conversation here been resolved, or not? @RalfJung, do you still think the call to `black_box` should be removed?

Comment (RalfJung) [reply src/test/run-pass/arbitrary_enum_discriminant.rs]: I am fine with keeping it.

File: src/test/ui/parser/tag-variant-disr-non-nullary.stderr
Comment (pnkfelix) [inline src/test/ui/parser/tag-variant-disr-non-nullary.stderr]: How hard would it be to continue to highlight the other variants with the assigned discriminator values?

(Or was it a deliberate choice here to limit attention to the first occurrence? If so, why?)

Comment (pnkfelix) [reply src/test/ui/parser/tag-variant-disr-non-nullary.stderr]: (Is this an artifact of invoking the feature gate diagnostic machinery?)

Comment (pnkfelix) [reply src/test/ui/parser/tag-variant-disr-non-nullary.stderr]: (it seems like your other tests did not run into a similar issue. Maybe I'll look more closely to figure out why this is occurring here.)

Comment (pnkfelix) [reply src/test/ui/parser/tag-variant-disr-non-nullary.stderr]: ah you explicitly noted this here: https://github.com/rust-lang/rust/pull/60732/files#r283095874

File: src/libsyntax/feature_gate.rs
Comment (pnkfelix) [reply src/libsyntax/feature_gate.rs]: Doesn't the output from your test [here](src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.stderr) contradict your claim here about the behavior of the `gate_feature` macros?

It certainly seems like that managed to emit the diagnostic and then continue.

What happens if you change the `if let Some(variant) = disr_units.next()` here into `for variant in disr_units` ?

**Comment by pnkfelix** [comment]
I have one minor [question](https://github.com/rust-lang/rust/pull/60732#discussion_r286467207) about the diagnostic code; once that is addressed, r=me.

**Comment by pnkfelix** [comment]
@jswrenn wrote:

> @oli-obk
> 
> > `unsafe { *(value as *const Self as *const u8) }` is only valid if you are using `repr(u8)`. This is mainly what I was referring to. Since the unsafe pointer magic is the only thing making it possible right now to extract he discriminant value, and it requires `repr(u8)`, I was wondering if we may as well require `repr(u8)` (or other ints) for enums with variants with fields and explicit discriminants.
> 
> I see your point. I think this is a good idea, and I'll take a stab at it in the next few days!
> 
> Implementation wise, I think [`rustc_typeck::check::check_enum`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/fn.check_enum.html) is the best place to add this check.

i don't mind landing the bulk of the code as is and landing the change suggested here in a later PR. I'll let @jswrenn decide which approach they would prefer to take.

**Comment by jswrenn** [comment]
I implemented the `repr` check last night (just need to update the tests), so I might as well push it with this PR!

File: src/libsyntax/feature_gate.rs
Comment (jswrenn) [reply src/libsyntax/feature_gate.rs]: It's emitting multiple "discriminants on non-unit variants are experimental" diagnostics, rather than one diagnostic with multiple span labels.

The old "discriminator values can only be used with a field-less enum" error message (which was implemented as a parsing error), emitted one diagnostic with multiple span labels (one for every variant with a discriminant). In the presence of the `arbitrary_enum_discriminant` feature, this is no longer an error (and in its absence, it should be an error). However, since the `gate_feature` macros immediately emit the diagnostic they generate, I cannot add span labels to it.

> What happens if you change the if let Some(variant) = disr_units.next() here into for variant in disr_units ?

Making this change would emit a distinct diagnostic for every unit variant with a discriminator. I don't really have any objections to doing this, it just causes quite a bit of clutter on stderr.

File: src/test/run-pass/arbitrary_enum_discriminant.rs
Comment (Centril) [inline src/test/run-pass/arbitrary_enum_discriminant.rs]: I would prefer to see all of these tests, including the feature gate tests, moved into `src/test/ui/enum-discriminant/*` for better organization and discoverability when it is time for stabilization.

File: src/test/run-pass/discriminant_value.rs
Comment (Centril) [inline src/test/run-pass/discriminant_value.rs]: It would be good to move this file to the folder aforementioned; Moving `discriminant_value-wrapper.rs` would also be good.

**Review by Centril — DISMISSED** [review]

File: src/test/run-pass/enum-discriminant/arbitrary_enum_discriminant.rs
Comment (Centril) [inline src/test/run-pass/enum-discriminant/arbitrary_enum_discriminant.rs]: Move `run-pass/enum-discriminant/*` -> `ui/enum-discriminant` and then put `// run-pass` in each of the moved files.

File: src/test/ui/arbitrary_enum_discriminant-no-repr.rs
Comment (Centril) [inline src/test/ui/arbitrary_enum_discriminant-no-repr.rs]: this

File: src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs
Comment (Centril) [inline src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs]: and this... should also be moved into `ui/enum-discriminant/` and then use `// gate-test-arbitrary_enum_discriminant` in this specific file.

Comment (jswrenn) [reply src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs]: Shall I move `test/ui/parser/tag-variant-disr-non-nullary.rs` too?

Comment (Centril) [reply src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs]: I think that one can remain as-is since the change there is mostly transitory and will be noticed when stabilization is done anyways.

Comment (jswrenn) [reply src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs]: > and this... should also be moved into `ui/enum-discriminant/` and then use `// gate-test-arbitrary_enum_discriminant` in this specific file.

The filename is already prefixed with `feature-gate-`. Is there a reason to prefer one convention for marking a file as a gate test over the other?

Comment (Centril) [reply src/test/ui/feature-gate/feature-gate-arbitrary_enum_discriminant.rs]: I assumed tidy wouldn't be happy otherwise; other than that idk.

**Comment by jswrenn** [comment]
Squashed to make the impending rebase on #61039 a little less laborious. Unfortunately, it'll be a few days before I can tackle this since I'm busy this weekend. As [mentioned](https://github.com/rust-lang/rust/pull/60732#discussion_r283095874), the convention of using `gate_feature` in feature_gate.rs makes it difficult to display rich diagnostics.

**Comment by jswrenn** [comment]
@Centril, @pnkfelix: I think everything's ready to go!

File: src/libsyntax/feature_gate.rs
Comment (jswrenn) [reply src/libsyntax/feature_gate.rs]: I think this is resolved. The emitted error looks like this:
https://github.com/rust-lang/rust/blob/d4d82af9cbb295d5c7da3ae44ff619e6e59a63c0/src/test/ui/enum-discriminant/feature-gate-arbitrary_enum_discriminant.stderr#L19-L32

**Comment by jswrenn** [comment]
@Centril, github still lists you as 'Requesting Changes', but the unresolved suggestion [isn't reachable](https://github.com/rust-lang/rust/pull/60732/files/315974942fa2a83377bf7d006f1e4d63b25b0c98#pullrequestreview-240682927) (because of a rebase), so I cannot mark it as resolved. To the best of my knowledge, I've resolved all issues raised so far.

File: src/libsyntax/feature_gate.rs
Comment (eddyb) [reply src/libsyntax/feature_gate.rs]: There are no fields though, the message seems misleading. Maybe say "tuple variant" and "struct variant"? Although when there are fields, it's better to say that.

Also, the spans reported on for the other variants shouldn't include the `= ...` part.

Comment (jswrenn) [reply src/libsyntax/feature_gate.rs]: > There are no fields though, the message seems misleading. Maybe say "tuple variant" and "struct variant"?

Agreed. The revised message is as follows:
https://github.com/rust-lang/rust/blob/9792356558a995c06d8a4152ebb2ed4ea2421f4e/src/test/ui/enum-discriminant/feature-gate-arbitrary_enum_discriminant.stderr#L19-L32

> Although when there are fields, it's better to say that.

I think this wording implies that the presence of fields is the issue, which is also misleading, so I stuck with "tuple variant" and "struct variant" unconditionally.

> Also, the spans reported on for the other variants shouldn't include the `= ...` part.

I agree, but I don't think there's an already-available span encapsulating a variant-definition-minus-the-discriminant. This error message's existence may be short-lived, so I don't know if it's worth expending too much effort to improve it.

**Comment by lucab** [comment]
(I think this may be waiting for another review round after updates, despite what the label says)

**Review by oli-obk — CHANGES_REQUESTED** [review]
just some stylistic nits, then this lgtm

File: src/libsyntax/feature_gate.rs
Comment (jswrenn) [reply src/libsyntax/feature_gate.rs]: Yes, I like this a lot!

**Pull request merged**
